### PR TITLE
Refactor verifier API

### PR DESCRIPTION
**Pull request by panicbit**
The most significant changes:
- `V1Token::verify -> V1Token::verify_integrity`
- `V1Token::verify` now takes a verifier to verify the token with
- Verifiers are types which implement the `Verifier` trait now

### CODE REVIEW

File: src/token.rs [hunk 1]
```diff
@@ -3,11 +3,13 @@ use std::marker::Sized;
...
+use verifier::Verifier;
 
 pub trait Token {
     fn new(key: &[u8], identifier: Vec<u8>, location: Option<Vec<u8>>) -> Self;
     fn deserialize(macaroon: Vec<u8>) -> Result<Self> where Self: Sized;
     fn serialize(&self) -> Result<Vec<u8>>;
     fn add_caveat(&self, caveat: &Caveat) -> Self;
-    fn verify(&self, key: &[u8]) -> Result<()>;
+    fn verify<V: Verifier>(&self, key: &[u8], verifier: V) -> Result<()>;
+    fn verify_integrity(&self, key: &[u8]) -> Result<()>;
```
Comment (tarcieri) [inline src/token.rs]: Perhaps this should be called `authenticate`

Comment (panicbit) [inline src/token.rs]: Yeah, that sounds good. I would still keep `verify_integrity` though, to avoid confusion between `authenticate` and `verify`.

**Comment by tarcieri** [comment]
Apologies for not responding on this PR. I think this is a good change. I wonder what can be done to better abstract first vs third party caveats. But that's a separate concern.

My only concerns are mostly around naming. I definitely appreciate clarity in cryptographic APIs, and longer-but-scarier names driving you to the common API, but I don't think that's what this is doing right now.

I would suggest `authenticate_without_verifying` and `verify` in that regard. The former checks cryptographic authenticity without checking caveats. The latter does both.

WDYT?

**Comment by panicbit** [comment]
I have mostly the same concerns. I'm not quite sure about `authenticate` since macaroons can also `authorize`, even without authenticating.

**Comment by tarcieri** [comment]
So as I see it, Macaroon verification happens in two steps:

* **AuthN**: we verify the Macaroon is authentic, via chained HMAC
* **AuthZ**: we verify the Macaroon is authorized, where the Macaroon being authentic is necessary but not sufficient. It must also pass all caveat checks.

I guess I still like `authenticate_without_verifying` and `verify` here, where the latter invokes the former. I am a fan of explicit cryptographic APIs which drive you to the API you should probably be using with brevity.

**Comment by panicbit** [comment]
I agree and made the changes.

**Comment by tarcieri** [comment]
Merged. This could probably use a `0.4.0` release.

**Comment by panicbit** [comment]
If you feel like it, yeah. [Semver doesn't require it though.](http://semver.org/#spec-item-4)

**Pull request merged**